server/src/main/java/org/elasticsearch/action/bulk/BulkItemResponse.java

```diff
@@ -341,66 +341,41 @@ public String toString() {
         }
     }
 
-    private int id;
+    private final int id;
 
-    private OpType opType;
+    private final OpType opType;
 
-    private DocWriteResponse response;
+    private final DocWriteResponse response;
 
-    private Failure failure;
-
-    BulkItemResponse() {}
+    private final Failure failure;
 
     BulkItemResponse(ShardId shardId, StreamInput in) throws IOException {
         id = in.readVInt();
         opType = OpType.fromId(in.readByte());
-
-        byte type = in.readByte();
-        if (type == 0) {
-            response = new IndexResponse(shardId, in);
-        } else if (type == 1) {
-            response = new DeleteResponse(shardId, in);
-        } else if (type == 3) { // make 3 instead of 2, because 2 is already in use for 'no responses'
-            response = new UpdateResponse(shardId, in);
-        } else if (type != 2) {
-            throw new IllegalArgumentException("Unexpected type [" + type + "]");
-        }
-
-        if (in.readBoolean()) {
-            failure = new Failure(in);
-        }
+        response = readResponse(shardId, in);
+        failure = in.readBoolean() ? new Failure(in) : null;
     }
 
     BulkItemResponse(StreamInput in) throws IOException {
         id = in.readVInt();
         opType = OpType.fromId(in.readByte());
-
-        byte type = in.readByte();
-        if (type == 0) {
-            response = new IndexResponse(in);
-        } else if (type == 1) {
-            response = new DeleteResponse(in);
-        } else if (type == 3) { // make 3 instead of 2, because 2 is already in use for 'no responses'
-            response = new UpdateResponse(in);
-        } else if (type != 2) {
-            throw new IllegalArgumentException("Unexpected type [" + type + "]");
-        }
-
-        if (in.readBoolean()) {
-            failure = new Failure(in);
-        }
+        response = readResponse(in);
+        failure = in.readBoolean() ? new Failure(in) : null;
     }
 
-    public BulkItemResponse(int id, OpType opType, DocWriteResponse response) {
+    private BulkItemResponse(int id, OpType opType, DocWriteResponse response, Failure failure) {
         this.id = id;
         this.response = response;
         this.opType = opType;
+        this.failure = failure;
```

Comment:
I think it makes sense to add an assertion here that either `response` or `failure` needs to be set, but not both.

Reply:
:+1: